include/rmm/cuda_event_view.hpp

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) 2020-2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <rmm/detail/error.hpp>
+
+#include <cuda_runtime_api.h>
+
+#include <atomic>
+#include <cstddef>
+#include <cstdint>
+
+namespace rmm {
+
+enum cuda_event_flags {
+  /** Default event flag. */
+  EVENT_DEFAULT = cudaEventDefault,
+  /** Event uses blocking synchronization. */
+  EVENT_BLOCKING_SYNC = cudaEventBlockingSync,
+  /** Event will not record timing data. */
+  EVENT_DISABLE_TIMING = cudaEventDisableTiming,
+  /** Event is suitable for interprocess use. cudaEventDisableTiming must be set. */
+  EVENT_INTERPROCESS = cudaEventInterprocess
+};
```

Comment:
Come to think of it, I wonder if these things should be encoded in the type of the `cuda_event/event_view`. For example, the `elapsed_time_since` function currently accepts any `event_view`, but really it only works with an event that was created where `cudaEventDisableTiming` wasn't specified. That's the kind of thing that the type system should be used to enforce.

Reply:
Hmm, I'm personally not sure if it's worth the cost. I think lifting `cudaEventDisableTiming` into something like `cuda_event_with_timing` would be nice. But the same thing would cause problems for `cuda_event_view`: we'd get the exception from `elapsed_time_since` moved to the implicit conversion/constructor from `cudaEvent_t`. `cudaEventBlockingSync` does not seem to affect the api at all, so there is no point in lifting it. Not sure if it makes sense for `cudaEventInterprocess` and the future flags to come either, and it may seem illogical to a user that the meaning of some of the flags is duplicated in the type system.

Reply:
I was thinking something more like `cuda_event<Properties...>`. 

So the API for `elapsed_time_since` would be something like:
```
template <typename... Properties>
float elapsed_time_since(event_view<Properties...> e){
   static_assert( /* Properties does not contain `cudaEventDisableTiming` */ )
}
```

> cudaEventBlockingSync does not seem to affect the api at all, so there is no point in lifting it

Maybe. Maybe not though. I might want to specialize/overload a function for an event that is blocking vs. non-blocking. 

Likewise with `cudaEventInterprocess`, I might write a function that I require the specified event is capable of working with IPC. I don't believe there is even a way to query a `cudaEvent_t` after the fact to detect if it was created with `cudaEventInterprocess`, so if a user passed in an incorrectly created stream, an error likely wouldn't be detected until I attempted to use that event in another API like `cudaIpcGetEventHandle`. 

My point is that there are a number of ways a user can misuse events. We can either detect those at runtime and throw exceptions, or we can detect them at compile time. Personally, I try and push as many errors to compile time as possible. 

Thinking through all the complexities and corner cases like this is exactly why we shouldn't have to be designing these things ourselves and should be provided by CUDA, but I digress...

Reply:
Can you convince me, @jrhemstad that there is never a need to decide the properties of an event at runtime?

Reply:
Nope, I definitely can't. I think what the right thing to do here is use a pattern like [`std::span`](https://en.cppreference.com/w/cpp/container/span).

A span can have a statically known size, like `std::span<int, 5>` is a span of 5 `int`s. 

Or it can have a dynamically known size `std::span<int>`. The way they do this is they have a special sentinel value for the second template argument that says "the size of this span is dynamic". It's called [`std::dynamic_extent`](https://en.cppreference.com/w/cpp/container/span/dynamic_extent) and it's the default value for the second template argument.

So we could use this same pattern for an `event` type where you could have:
```
namespace cuda_event{
struct timing{};
struct IPC{};
struct dynamic{
   cudaEventFlags f;
   dynamic(cudaEventFlags f) : f{f}
};
}

cuda_event<timing> e0; // This is an event where timing hasn't been disabled
cuda_event<timing, IPC> e1; // Event that supports timing and IPC
cuda_event<dynamic> e2{/*dynamic value*/};
```

This way you can have best of both worlds. If I'm writing a function and I want to statically declare that an event passed into my function supports timing then I can specify `cuda_event<timing,...>`. If I want to support an event with dynamic properties I can just do `cuda_event<dynamic>`.

Don't get me wrong, it's a lot of machinery to make this work, but that's precisely why CUDA should be providing it :)

Reply:
I think having variable number of template arguments would be not ideal, because this would imply the markers can go in any order, and users later would have problems with `cuda_event<timing, IPC>` and `cuda_event<IPC,timing>` being different types. Perhaps, I can force the order and other constraints using lots of `static_assert`s...

In general, I'm always up for more type safety. It's just feels to me a bit foreign in rmm, where there are no similar constructs on the streams (legacy/default/(non-)blocking) and stream pools (non-empty?:).

Reply:
How about this?
https://github.com/rapidsai/rmm/blob/9aadcb43ec9a3866fc87ef79ed1ee382f33d9345/include/rmm/cuda_event_view.hpp#L82-L83
...
https://github.com/rapidsai/rmm/blob/9aadcb43ec9a3866fc87ef79ed1ee382f33d9345/include/rmm/cuda_event_view.hpp#L162-L163

This way, `cuda_event_view` (without brackets) defaults to the most commonly used, statically enforced `cuda_event_view<EVENT_DEFAULT>`, while it's still possible to set flags dynamically via `cuda_event_view_`. And since I use `cuda_event_flags` as the template argument, I can use union (`|`) operator and not worry about the order in which the flags appear.